blockchain/internal/spendpruner/pruner.go

```diff
@@ -0,0 +1,263 @@
+		if consumerId == id {
+			return true
+		}
+	}
+
+	return false
+}
+
+// AddSpendConsumerDeps adds an entry for each spend consumer dependent on
+// journal data for the provided block hash.
+func (s *SpendJournalPruner) AddSpendConsumerDeps(blockHash *chainhash.Hash) error {
+	s.consumersMtx.RLock()
+	defer s.consumersMtx.RUnlock()
+
+	for _, consumer := range s.consumers {
+		if s.dependencyExistsInternal(blockHash, consumer.ID()) {
+			// Dependency already created, skip.
+			continue
+		}
+
+		needSpendData, err := consumer.NeedSpendData(blockHash)
+		if err != nil {
+			msg := fmt.Sprintf("unable to assert dependency: %s", err)
+			return pruneError(ErrNeedSpendData, msg)
+		}
+
+		if !needSpendData {
+			continue
+		}
+
+		// Add a spend dependency entry of the block hash for the consumer.
+		s.dependentsMtx.Lock()
+		dependents, ok := s.dependents[*blockHash]
+		if !ok {
+			dependents = []string{consumer.ID()}
+			s.dependents[*blockHash] = dependents
+			s.dependentsMtx.Unlock()
+
+			continue
+		}
+
+		dependents = append(dependents, consumer.ID())
+		s.dependents[*blockHash] = dependents
+		s.dependentsMtx.Unlock()
+
+	}
+
+	s.dependentsMtx.Lock()
+	dependents := s.dependents[*blockHash]
+	s.dependentsMtx.Unlock()
+
+	// Update the persisted spend consumer deps entry for
+	// the provided block hash.
+	err := s.db.Update(func(tx database.Tx) error {
+		err := dbUpdateSpendConsumerDeps(tx, *blockHash, dependents)
+		return err
+	})
+	if err != nil {
+		return fmt.Errorf("unable to update persisted consumer "+
+			"dependencies for block hash %v: %v", blockHash, err)
+	}
+
+	return nil
+}
+
+// RemoveSpendConsumerDependency removes the provided spend consumer dependency
+// associated with the provided block hash from the spend pruner. The block
+// hash is removed as a key of the dependents map once all its dependency
+// entries are removed.
+func (s *SpendJournalPruner) RemoveSpendConsumerDependency(blockHash *chainhash.Hash, consumerID string) {
+	s.dependentsMtx.Lock()
+	defer s.dependentsMtx.Unlock()
+
+	dependents, ok := s.dependents[*blockHash]
+	if !ok {
+		// No entry for block hash found, do nothing.
+		return
+	}
+
+	for idx := 0; idx < len(dependents); idx++ {
+		if dependents[idx] == consumerID {
+			dependents = append(dependents[:idx], dependents[idx+1:]...)
+			s.dependents[*blockHash] = dependents
+			break
+		}
+	}
+
+	if len(dependents) == 0 {
+		delete(s.dependents, *blockHash)
+		s.pruneCh <- blockHash
+	}
+}
+
+// removeSpendConsumerDeps removes the key/value pair of spend consumer deps and
+// the provided block hash from the the prune set.
+func (s *SpendJournalPruner) removeSpendConsumerDeps(blockHash *chainhash.Hash) {
+	s.dependentsMtx.Lock()
+	defer s.dependentsMtx.Unlock()
+
+	delete(s.dependents, *blockHash)
+
+	// Remove the tracked spend journal entry for the provided
+	// block hash.
+	err := s.db.Update(func(tx database.Tx) error {
+		return dbUpdateSpendConsumerDeps(tx, *blockHash, nil)
+	})
```

Comment:
Is it intended that in this case, the db is updated after deleting `blockHash` from `s.dependents`, but the db is not updated in `RemoveSpendConsumerDependency`?  It appears that it might be useful for `RemoveSpendConsumerDependency` to use `removeSpendConsumerDeps` so that the removal logic is in one place.

Reply:
Good catch, `RemoveSpendConsumerDependency` needs to persist to the db after the removal. `removeSpendConsumerDeps` wipes all the spend consumer deps for the provided blockhash and updates the db when done while `RemoveSpendConsumerDependency` only wipes the spend consumer set for the provided blockhash on condition the last entry was just removed. For that reason I don't think calling `removeSpendConsumerDeps` from `RemoveSpendConsumerDependency` to keep removal logic in one place would be ideal since `removeSpendConsumerDeps` would have to be updated to cover the two contexts described.